src/files/files.ts

```diff
@@ -0,0 +1,250 @@
+		collab: Collab,
+		types: string[],
+	) {
+		this.collab = collab
+		this.excalidrawApi = excalidrawApi
+		this.types = types
+		const containerRef = document.getElementsByClassName(
+			'excalidraw-container',
+		)[0]
+		const constructedFile: BinaryFileData = {
+			mimeType: 'image/png',
+			created: 0o0,
+			id: 'placeholder_image' as FileId,
+			dataURL: 'data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHZpZXdCb3g9IjAgMCAyNCAyNCI+PHBhdGggZD0iTTE0LDJMMjAsOFYyMEEyLDIgMCAwLDEgMTgsMjJINkEyLDIgMCAwLDEgNCwyMFY0QTIsMiAwIDAsMSA2LDJIMTRNMTgsMjBWOUgxM1Y0SDZWMjBIMThNMTIsMTlMOCwxNUgxMC41VjEySDEzLjVWMTVIMTZMMTIsMTlaIiAvPjwvc3ZnPg==' as DataURL,
+		}
+		this.collab.addFile(constructedFile)
+		if (containerRef) {
+			containerRef.addEventListener('drop', (ev) =>
+				this.filesDragEventListener(ev),
+			)
+		}
+		let lastPointerDown = 0
+		// eslint-disable-next-line @typescript-eslint/no-unused-vars
+		this.excalidrawApi.onPointerDown((activeTool, state, event) => {
+			const clickedElement = state.hit.element
+			if (!clickedElement || !clickedElement.customData) {
+				return
+			}
+			event.stopPropagation()
+			if (Date.now() - lastPointerDown > 200) {
+				lastPointerDown = Date.now()
+				return
+			} else {
+				lastPointerDown = Date.now()
+			}
+			this.downloadFile(clickedElement.customData.meta)
+		})
+	}
+
+	private downloadFile(meta: Meta) {
+		const file = this.excalidrawApi.getFiles()[meta.fileId]
+		const blob = new Blob([file.dataURL], { type: meta.type })
+		const url = URL.createObjectURL(blob)
+		const a = document.createElement('a')
+		a.href = url
+		a.download = meta.name
+		a.click()
+		URL.revokeObjectURL(url)
+	}
+
+	private filesDragEventListener(ev: Event) {
+		if (ev instanceof DragEvent) {
+			for (const file of Array.from(ev.dataTransfer?.files || [])) {
+				this.handleFileInsert(file, ev)
+			}
+		}
+	}
+
+	private handleFileInsert(file: File, ev: Event) {
+		// if excalidraw can handle it, do nothing
+		if (this.types.includes(file.type)) {
+			return
+		}
+		ev.stopImmediatePropagation()
+
+		const fr = new FileReader()
+		fr.readAsDataURL(file)
+		fr.onload = () => {
+			const constructedFile: BinaryFileData = {
+				mimeType: file.type,
+				created: Date.now(),
+				id: (Math.random() + 1).toString(36).substring(7) as FileId,
+				dataURL: fr.result as DataURL,
+			}
+			if (typeof fr.result === 'string') {
+				const meta: Meta = {
+					name: file.name,
+					type: file.type,
+					lastModified: file.lastModified,
+					dataurl: fr.result,
+					fileId: constructedFile.id,
+				}
+				this.addCustomFileElement(constructedFile, meta)
+			}
+		}
+	}
+
+	private addCustomFileElement(constructedFile: BinaryFileData, meta: Meta) {
+		this.collab.portal.sendImageFiles({ [constructedFile.id]: constructedFile })
+		const elements = this.excalidrawApi
+			.getSceneElementsIncludingDeleted()
+			.slice()
+		const newElements = convertToExcalidrawElements([
```

Comment:
For the mime type icon you can use `window.OC.MimeType.getIconUrl(mimetype)` to get the URL of the svg file, guess we could fetch that and then insert it into the whiteboard